### ISSUE TITLE
Where is the camera control in the code, currently the demo is in the default gradio and the camera pans from left to right, thank you

### ISSUE THREAD

**Issue by D-Mad** [zqh0253/3DitScene#5]
Where is the camera control in the code, currently the demo is in the default gradio and the camera pans from left to right, thank you

![image](https://github.com/zqh0253/3DitScene/assets/8207507/5ac220f2-4207-4cfa-bfd7-5a3d7912172a)
How can I control the camera and objects?

**Comment by zqh0253** [comment]
The data module is defined in [this file](https://github.com/zqh0253/3DitScene/blob/main/threestudio/data/uncond.py) and initialized [here](https://github.com/zqh0253/3DitScene/blob/d9c4592cbca692d939ef635b36b46e603ff3500f/launch.py#L168).

For example, if you want to modify the inference camera trajectory, you can modify the `RandomCameraDataset::__init__` function at [here](https://github.com/zqh0253/3DitScene/blob/d9c4592cbca692d939ef635b36b46e603ff3500f/threestudio/data/uncond.py#L370).

**Comment by ThePassedWind** [comment]
Great work!
Here, I add another question.
How to get 3D mask by language feature distillation?

**Comment by zqh0253** [comment]
> Great work! Here, I add another question. How to get 3D mask by language feature distillation?

Thanks for your kind words. `GaussianBaseModel` has a variable called `ooi_masks_0` which stores the 3D mask of the object. 

For example, we use the 3D mask to select the object and perform rotation at [here](https://github.com/zqh0253/3DitScene/blob/d9c4592cbca692d939ef635b36b46e603ff3500f/custom/threestudio-3dgs/geometry/gaussian_base.py#L876C29-L876C40).

**Comment by ThePassedWind** [comment]
> > Great work! Here, I add another question. How to get 3D mask by language feature distillation?
> 
> Thanks for your kind words. `GaussianBaseModel` has a variable called `ooi_masks_0` which stores the 3D mask of the object.
> 
> For example, we use the 3D mask to select the object and perform rotation at [here](https://github.com/zqh0253/3DitScene/blob/d9c4592cbca692d939ef635b36b46e603ff3500f/custom/threestudio-3dgs/geometry/gaussian_base.py#L876C29-L876C40).

Thanks for your reply. Your code is highly implemented.
But I'm curious that are there details for implementation of other editing tasks such as moving and removal.
And the details for layout augmentation.

**Comment by zqh0253** [comment]
> > > Great work! Here, I add another question. How to get 3D mask by language feature distillation?
> > 
> > 
> > Thanks for your kind words. `GaussianBaseModel` has a variable called `ooi_masks_0` which stores the 3D mask of the object.
> > For example, we use the 3D mask to select the object and perform rotation at [here](https://github.com/zqh0253/3DitScene/blob/d9c4592cbca692d939ef635b36b46e603ff3500f/custom/threestudio-3dgs/geometry/gaussian_base.py#L876C29-L876C40).
> 
> Thanks for your reply. Your code is highly implemented. But I'm curious that are there details for implementation of other editing tasks such as moving and removal. And the details for layout augmentation.

You can change `GaussianBaseModel`'s property `_xyz` to achieve moving, and `_opacity_mask` to achieve removal. You can find example for moving [here](https://github.com/zqh0253/3DitScene/blob/d9c4592cbca692d939ef635b36b46e603ff3500f/gradio_app_single_process.py#L530), and removal [here](https://github.com/zqh0253/3DitScene/blob/d9c4592cbca692d939ef635b36b46e603ff3500f/custom/threestudio-3dgs/system/scene_lang.py#L115).

**Comment by ThePassedWind** [comment]
Thanks!
What about the implementation of layout augmentation.
A kind of implementation is to compute a convex hull to obtain the whole 3d mask.

**Comment by D-Mad** [comment]
> The data module is defined in [this file](https://github.com/zqh0253/3DitScene/blob/main/threestudio/data/uncond.py) and initialized [here](https://github.com/zqh0253/3DitScene/blob/d9c4592cbca692d939ef635b36b46e603ff3500f/launch.py#L168).
> 
> For example, if you want to modify the inference camera trajectory, you can modify the `RandomCameraDataset::__init__` function at [here](https://github.com/zqh0253/3DitScene/blob/d9c4592cbca692d939ef635b36b46e603ff3500f/threestudio/data/uncond.py#L370).

Thank you very much, can I ask for the code you edited to have the movement of cherry.gif?
https://github.com/zqh0253/3DitScene/blob/d9c4592cbca692d939ef635b36b46e603ff3500f/assets/cherry.gif

**Comment by zqh0253** [comment]
> > The data module is defined in [this file](https://github.com/zqh0253/3DitScene/blob/main/threestudio/data/uncond.py) and initialized [here](https://github.com/zqh0253/3DitScene/blob/d9c4592cbca692d939ef635b36b46e603ff3500f/launch.py#L168).
> > For example, if you want to modify the inference camera trajectory, you can modify the `RandomCameraDataset::__init__` function at [here](https://github.com/zqh0253/3DitScene/blob/d9c4592cbca692d939ef635b36b46e603ff3500f/threestudio/data/uncond.py#L370).
> 
> Thank you very much, can I ask for the code you edited to have the movement of cherry.gif? https://github.com/zqh0253/3DitScene/blob/d9c4592cbca692d939ef635b36b46e603ff3500f/assets/cherry.gif

The dataset has `rotate_around` function which defines the camera trajectory in the cherry example. Here is an example how to use this trajectory:

```
dm = threestudio.find(cfg.data_type)(cfg.data)
dm.setup(stage='test')
rot_data = dict()
frame_n = 120
for i in trange(frame_n):
    data = dm.test_dataset.rotate_around(60, i/frame_n*2*np.pi, scale=0.2)
    for k in data.keys():
        try:
            data[k] = torch.tensor([data[k]])
            data[k] = data[k].cuda()
        except:
            pass
        if k in ['c2w']:
            data[k] = data[k].unsqueeze(0)
    rot_data[i] = data

with torch.no_grad():
    video_writer = imageio.get_writer(name, fps=60)

    idx_list = list(range(frame_n)) + [0]
    for i in tqdm(idx_list):
        data = rot_data[i]
        res = system(data)
        rgb = res['comp_rgb'][0].cpu().numpy()
        rgb = (rgb * 255).astype(np.uint8)    
        video_writer.append_data(rgb)
    
    video_writer.close()
```

**Comment by zqh0253** [comment]
> Thanks! What about the implementation of layout augmentation. A kind of implementation is to compute a convex hull to obtain the whole 3d mask.

The augmentation uses ooi_mask_0 to identify which Gaussians belong to the object, and then applies translation, rotation, or removal to those Gaussians.

**Comment by ThePassedWind** [comment]
> > Thanks! What about the implementation of layout augmentation. A kind of implementation is to compute a convex hull to obtain the whole 3d mask.
> 
> The augmentation uses ooi_mask_0 to identify which Gaussians belong to the object, and then applies translation, rotation, or removal to those Gaussians.



> > Thanks! What about the implementation of layout augmentation. A kind of implementation is to compute a convex hull to obtain the whole 3d mask.
> 
> The augmentation uses ooi_mask_0 to identify which Gaussians belong to the object, and then applies translation, rotation, or removal to those Gaussians.

Thank you very much!
